README.md
SSDP Responder for Linux/UNIX
=============================
[![License Badge][]][License] [![GitHub Status][]][GitHub] [![Coverity Status][]][Coverity Scan]

Simple Service Discovery Protocol daemon (SSDP) for networked Linux and
UNIX devices.  Useful in any setup, big or small, but targeted more at
embedded systems that need to announce themselves to Windows systems.

`ssdpd` is a stand-alone UNIX daemon with no external dependencies but
the standard C library.  It has a built-in web server for serving the
UPnP XML description which Windows use to present the icon, by default
an InternetGatewayDevice is announced.

Also included is the `ssdp-scan` tool, which continuously scans for
SSDP capable hosts on the network.  Take care only to use this for
debugging since it scans the network quite aggressively.

> **Note:** when `ssdpd` is started as root it will drop privileges as
> soon as it has created its UUID cache and PID file.  It is recommended
> to have a dedicated `ssdp` user (and group) available in the system
> for this purpose.  As a fallback, user `nobody` (and `nogroup`) is
> used.  When started as non-root, make sure to provide a path for the
> cache file with the `-c FILE` option.


Usage
-----

```
Usage: ssdpd [-hnsvw] [-c FILE] [-d URL] [-i SEC] [-l LEVEL] [-m NAME] [-M URL]
                      [-p URL] [-P FILE] [-r SEC] [-R NUM] [-t TTL] [-u UUID]
                      [IFACE [IFACE ...]]

    -c FILE   Path to alternate ssdpd.cache to store and/or read the UUID
    -d URL    Override UPnP description.xml URL in announcements.  The '%s' in
              the URL is replaced with the IP, e.g. https://%s:1901/main.xml
    -h        This help text
    -i SEC    SSDP notify interval (30-900), default 300 sec
    -l LVL    Set log level: none, err, notice (default), info, debug
    -m NAME   Override manufacturer in the default description.xml
    -M URL    Override manufacturerURL in the default description.xml
    -n        Run in foreground, do not daemonize by default
    -r SEC    Interface refresh interval (5-1800), default 600 sec
    -R NUM    Initial retries, using 10 sec refresh interval, default 3 times
    -p URL    Override presentationURL (WebUI) in the default description.xml
              The '%s' is replaced with the IP address.  Default: http://%s/
    -P FILE   Override PID file location, absolute path required
    -s        Use syslog, default unless running in foreground, -n
    -t TTL    TTL for multicast frames, default 2, according to the UDA
    -u UUID   Custom UUID instead of auto-generating one
    -v        Show program version
    -w        Disable built-in micro HTTP server on port 1901

Bug report address : https://github.com/troglobit/ssdp-responder/issues
Project homepage   : https://github.com/troglobit/ssdp-responder
```

The `-d URL` argument can contain one `%s` modifier which is replaced
with the IP address of the interface the SSDP notification or reply is
sent on.  For example:

    ssdpd -d https://%s:1901/description.xml

See `configure --help` for some build time options.


Example
-------

The following example assumes the system `eth0` interface is connected
to an ISP and `eth1` to the LAN.  Every 300 sec the list of addresses
for `eth1` are updated, if a new address is added a `NOTIFY *` message
is sent, otherwise `NOTIFY *` messages are sent every 30 seconds.

```
ssdpd -i 30 -r 300 eth1
```


Configure & Build
-----------------

The GNU Configure & Build system use `/usr/local` as the default install
prefix.  In many cases this is useful, but this means the configuration
files, cache, and PID files will also use that prefix.  Most users have
come to expect those files in `/etc/` and `/var/` and configure has a
few useful options that are recommended to use.  Hence, you may want to
use something like this:

    ./configure --prefix=/usr --sysconfdir=/etc --runstatedir=/var/run
    make -j$(($(nproc) + 1))
    sudo make install-strip

Usually your system reserves `/usr` for native pacakges, so most users
drop `--prefix`, installing to `/usr/local`, or use `--prefix=/opt`.

**Note:** On some systems `--runstatedir` may not be available in the
  configure script, try `--localstatedir=/var` instead.


### Building from GIT

The `configure` script and the `Makefile.in` files are generated for
release tarballs and not stored in GIT.  When you work with the GIT
source tree you need the GNU `automake` and `autoconf` tools:

    $ sudo apt install automake autoconf

Now, from the top directory of the cloned GIT tree, call:

    $ ./autogen.sh


### Static Build

Some people want to build statically, to do this with `autoconf` add the
following `LDFLAGS=` *after* the configure script.  You may also need to
add `LIBS=...`, which will depend on your particular system:

    ./configure LDFLAGS="-static" ...


Integration with systemd
------------------------

For systemd integration `libsystemd-dev` and `pkg-config` are required.
When the unit file is installed, `systemctl` can be used to enable and
start the daemon:

    $ sudo systemctl enable ssdpd.service
    $ sudo systemctl start  ssdpd.service

Check that it started properly by inspecting the system log, or:

    $ sudo systemctl status ssdpd.service

To stop the service:

    $ sudo systemctl stop   ssdpd.service


Origin
------

Cloned from [mrdisc](https://github.com/troglobit/mrdisc) and whacked at
with a bat until it replies to SSDP "MSEARCH *" messages used by Windows.

[License]:         https://en.wikipedia.org/wiki/ISC_license
[License Badge]:   https://img.shields.io/badge/License-ISC-blue.svg
[GitHub]:          https://github.com/troglobit/ssdp-responder/actions/workflows/build.yml/
[GitHub Status]:   https://github.com/troglobit/ssdp-responder/actions/workflows/build.yml/badge.svg
[Coverity Scan]:   https://scan.coverity.com/projects/20496
[Coverity Status]: https://scan.coverity.com/projects/20496/badge.svg


ChangeLog.md
Change Log
==========

All notable changes to the project are documented in this file.


[v2.1][] - 2023-06-03
---------------------

### Changes
 - Add support for JSON output with `ssdpd-scan -j`, issue #17
 - Dump received data from devices in `ssdp-scan -l debug`
 - Termios support for disabling `^C` output from `ssdp-scan`
 - Update copyright years, LICENSE file hash change
 - Source `/etc/default/ssdpd` from systemd unit file

### Fixes
 - Fix #15: add fallback to `/var/lib/ssdpd/uuid.cache`
 - Fix #16: truncated UUID when reading from cache file


[v2.0][] - 2023-02-19
---------------------

The drop privs release.

### Changes
 - Add support for `-c NAME` to override `ssdpd` UUID cache file
 - Add support for `-d URL` to override the UPnP description URL, a
   single `%s` is supported to be replaced with the interface address
 - Add support for `-m NAME` to override the `manufacturer` in the
   default `description.xml`
 - Add support for `-M URL` to override the `manufacturerURL` in the
   default `description.xml`
 - Add support for `-p URL` to override the `presentationURL` in the
   default `description.xml`, a single `%s` is supported
 - Add support for `-P FILE` to override the default PID file location
   default `description.xml`, a single `%s` is supported
 - Add support for `-u UUID` to use a custom UUID, useful when the
   built-in micro HTTP server is disabled
 - Add support for `-w` to disable built-in micro HTTP server, useful
   when other, more capable, web servers are available.  Make sure to
   have the alternate web server running on port 1901 to serve the file
   `/description.xml`, see also `-d URL` above, which details the
   location of the UPnP description URL
 - Increased debug messages in `ssdp_init()`, for troubleshooting
 - Add support for dropping root privileges after initial setup.  The
   used `ssdp` is searched for first, with fallback to `nobody`
 - Add support for an initial retry interval at startup, in case ssdpd
   is started before any interface has been configured.  Configurable
   retry count using `-R NUM` command line option
 - Add support for an IP address monitor using Linux netlink interfaces
   coming and going at runtime.  Meaning, `-R NUM` is unused on Linux.
   For this to work, `ssdpd` must start as root or have `CAP_NET_RAW`

### Fixes
 - Fix #11: periodic busy loop causing intermittent 100% CPU load
 - Fix invalid `<UDN>uuid:uuid:...</UDN>` in `description.xml`
 - Add `Date:` and `Server:` to HTTP header in micro HTTP server
 - Add support for HTTP HEAD requests to micro HTTP server
 - Handle case when `/etc/os-release` does not have `VERSION_ID` fall
   back to use `VERSION_CODENAME`
 - Don't overwrite CPPFLAGS from the command line
 - Portability fix to `utimensat()` replacement function


[v1.9][] - 2022-10-30
---------------------

The [Dennis Ritchie](https://www.oreilly.com/content/dennis-ritchie-day/) release.

### Changes
 - Add command line options to `sscp-scan`, mostly for testing but may
   be useful for other purposes too
 - Use `$ac_default_prefix` instead of `/usr/local` in configure script
   when expanding paths
 - Update copyright years (affects LICENSE file hash)
 - Add ChangeLog to project
 - Drop Travis-CI in favor of GitHub Actions

### Fixes
 - Fix #6: workaround for OpenVPN /32 default server setup
 - Fix #9: time-of-check vs time-of-use issue with caching of UUID,
   found by Coverity Scan, fixed by Raul Porancea
 - Fix #10: basic instructions for building and starting the daemon


[v1.8][] - 2021-01-23
---------------------

### Changes
 - Use UUID cache directory from configure script, with fall-back to
   operating system specific `/var/lib/misc` or `/var/db`
 - Update man page with info on UUID cache location
 - Add missing systemd unit file
 - Add missing SysV init script

### Fixes
 - Fix avahi-daemon (mDNS) conflict, no conflict with SSDP, caused by
   copy-paste between mdnsd and project and this
 - Fix default install prefix, should be GNU `/usr/local` not `/`.  The
   default for Debian systems is `/usr`
 - Workaround for `--runstatedir` on systems with older autoconf


[v1.7][] - 2020-06-07
---------------------

### Changes
 - Use `/etc/os-release` as base for SSDP server string, distribution
   release information is primarily stored in this file, some Linux
   systems still use `/etc/lsb-release`, on real UNIX systems we can use
   `uname(1)`
 - Update [ssdpd(8)](https://man.troglobit.com/man8/ssdpd.8.html) man page

### Fixes
 - N/A


[v1.6][] - 2020-06-07
---------------------

### Changes
 - Refactor socket handling, enable `SO_REUSEADDR` + `SO_REUSEPORT`
 - Change default behavior, must now use `-n` to run in foreground
 - Add `ssdp-scan` tool, similar to `mdns-scan`
 - Add systemd unit file
 - Initial Debian/Ubuntu packaging

### Fixes
 - Fix #1: CVE-2019-14323


[v1.5][] - 2017-09-23
---------------------

### Changes
 - New default SSDP notify interval, 300 sec. (Was 600 sec)
 - Validation of refresh and notify intervals, the notify interval must
   not be longer than half the cache timeout
   - Notify: 30-900 sec
   - Refresh: 5-1800 sec

### Fixes
 - Fix regression in v1.3, allow running without interface filtering
 - The new interface filtering feature of v1.4 managed to filter out all
   interfaces if none were given on the command line.
 - Fix CPU overload problem, caused by invalid timer comparison


[v1.4][] - 2017-05-16
---------------------

Minor bug fix release.

### Fixes
 - Fix invalid argument to `accept()` in web server


[v1.3][] - 2017-05-16
---------------------

### Changes
 - Massive refactor
 - Support SSDP per interface and multiple addresses per interface
 - Use `SOCK_DGRAM` instead of `SOCK_RAW`, no longer need to run as root
 - New refresh timer, runs independently of NOTIFY timer, checks for new addresses


[v1.2][] - 2017-05-16
---------------------

Announces itself as an InternetGatewayDevice, works with Windows, serves
XML description.

### Changes

src/ssdpd.c
/* SSDP responder
 *
 * Copyright (c) 2017-2023  Joachim Wiberg <troglobit@gmail.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.a
 */

#include <sys/utsname.h>	/* uname() for !__linux__ */
#include <sys/stat.h>
#include <sys/types.h>
#include <pwd.h>

#include "ssdp.h"

char  server_string[64] = "POSIX UPnP/1.0 " PACKAGE_NAME "/" PACKAGE_VERSION;
char  location[128];
char  hostname[64];
int   ttl = MC_TTL_DEFAULT;
char *cachefn = NULL;
char *ver = NULL;
char *os  = NULL;
#ifdef MANUFACTURER_URL
char  mfrurl[128] = MANUFACTURER_URL;
#else
char  mfrurl[128];
#endif
char  mfrnm[128] = MANUFACTURER;
char  uuid[42];
char  url[128] = "http://%s/";

char **ifs;
size_t ifnum;

static char *supported_types[] = {
	SSDP_ST_ALL,
	"upnp:rootdevice",
	"urn:schemas-upnp-org:device:InternetGatewayDevice:1",
	uuid,
	NULL
};

volatile sig_atomic_t running = 1;
volatile sig_atomic_t recheck = 1;

extern void web_init(void);


static void compose_addr(struct sockaddr_in *sin, char *group, int port)
{
	memset(sin, 0, sizeof(*sin));
	sin->sin_family      = AF_INET;
	sin->sin_port        = htons(port);
	sin->sin_addr.s_addr = inet_addr(group);
}

static char *compose_location(char *host)
{
	const char *fmt = location;
	static char buf[256];

	snprintf(buf, sizeof(buf), fmt, host);
	return buf;
}

static void compose_response(char *type, char *host, char *buf, size_t len)
{
	char usn[256];
	char date[42];
	time_t now;

	/* RFC1123 date, as specified in RFC2616 */
	now = time(NULL);
	strftime(date, sizeof(date), "%a, %d %b %Y %T %Z", gmtime(&now));

	if (type) {
		if (!strcmp(type, uuid))
			type = NULL;
		else
			snprintf(usn, sizeof(usn), "%s::%s", uuid, type);
	}

	if (!type)
		strlcpy(usn, uuid, sizeof(usn));

	snprintf(buf, len, "HTTP/1.1 200 OK\r\n"
		 "Server: %s\r\n"
		 "Date: %s\r\n"
		 "Location: %s\r\n"
		 "ST: %s\r\n"
		 "EXT: \r\n"
		 "USN: %s\r\n"
		 "Cache-Control: max-age=%d\r\n"
		 "\r\n",
		 server_string,
		 date,
		 compose_location(host),
		 type,
		 usn,
		 CACHE_TIMEOUT);
}

static void compose_notify(char *type, char *host, char *buf, size_t len)
{
	char usn[256];

	if (type) {
		if (!strcmp(type, SSDP_ST_ALL))
			type = NULL;
		else
			snprintf(usn, sizeof(usn), "%s::%s", uuid, type);
	}

	if (!type) {
		type = usn;
		strlcpy(usn, uuid, sizeof(usn));
	}

	snprintf(buf, len, "NOTIFY * HTTP/1.1\r\n"
		 "Host: %s:%d\r\n"
		 "Server: %s\r\n"
		 "Location: %s\r\n"
		 "NT: %s\r\n"
		 "NTS: ssdp:alive\r\n"
		 "USN: %s\r\n"
		 "Cache-Control: max-age=%d\r\n"
		 "\r\n",
		 MC_SSDP_GROUP, MC_SSDP_PORT,
		 server_string,
		 compose_location(host),
		 type,
		 usn,
		 CACHE_TIMEOUT);
}

size_t pktlen(unsigned char *buf)
{
	size_t hdr = sizeof(struct udphdr);

	return strlen((char *)buf + hdr) + hdr;
}

static void send_message(struct ifsock *ifs, char *type, struct sockaddr *sa, socklen_t salen)
{
	struct sockaddr_in dest;
	char host[NI_MAXHOST];
	char buf[MAX_PKT_SIZE];
	size_t note = 0;
	ssize_t num;
	int s;

	if (ifs->addr.sin_addr.s_addr == htonl(INADDR_ANY))
		return;

	gethostname(hostname, sizeof(hostname));
	s = getnameinfo((struct sockaddr *)&ifs->addr, sizeof(struct sockaddr_in), host, sizeof(host), NULL, 0, NI_NUMERICHOST);
	if (s) {
		logit(LOG_WARNING, "Failed getnameinfo(): %s", gai_strerror(s));
		return;
	}

	if (ifs->addr.sin_addr.s_addr == htonl(INADDR_ANY))
		return;

	if (!strcmp(type, SSDP_ST_ALL))
		type = NULL;

	memset(buf, 0, sizeof(buf));
	if (sa)
		compose_response(type, host, buf, sizeof(buf));
	else
		compose_notify(type, host, buf, sizeof(buf));

	if (!sa) {
		note = 1;
		compose_addr(&dest, MC_SSDP_GROUP, MC_SSDP_PORT);
		sa = (struct sockaddr *)&dest;
		salen = sizeof(struct sockaddr_in);
	}

	logit(LOG_DEBUG, "Sending %s from %s ...", !note ? "reply" : "notify", host);
	num = sendto(ifs->sd, buf, strlen(buf), 0, sa, salen);
	if (num < 0)
		logit(LOG_WARNING, "Failed sending SSDP %s, type: %s: %s", !note ? "reply" : "notify", type, strerror(errno));
}

static void ssdp_recv(int sd)
{
	char buf[MAX_PKT_SIZE + 1];
	struct sockaddr sa;
	socklen_t salen;
	ssize_t len;

	memset(buf, 0, sizeof(buf));
	salen = sizeof(sa);
	len = recvfrom(sd, buf, sizeof(buf) - 1, MSG_DONTWAIT, &sa, &salen);
	if (len > 0) {
		buf[len] = 0;

		if (sa.sa_family != AF_INET)
			return;

		if (strstr(buf, "M-SEARCH *")) {
			struct sockaddr_in *sin = (struct sockaddr_in *)&sa;
			struct ifsock *ifs;
			char *ptr, *type;
			size_t i;

			ifs = ssdp_find(&sa);
			if (!ifs) {
				logit(LOG_DEBUG, "No matching socket for client %s", inet_ntoa(sin->sin_addr));
				return;
			}
			logit(LOG_DEBUG, "Matching socket for client %s", inet_ntoa(sin->sin_addr));

			type = strcasestr(buf, "\r\nST:");
			if (!type) {
				logit(LOG_DEBUG, "No Search Type (ST:) found in M-SEARCH *, assuming " SSDP_ST_ALL);
				type = SSDP_ST_ALL;
				send_message(ifs, type, &sa, salen);
				return;
			}

			type = strchr(type, ':');
			if (!type)
				return;
			type++;
			while (isspace(*type))
				type++;

			ptr = strstr(type, "\r\n");
			if (!ptr)
				return;
			*ptr = 0;

			for (i = 0; supported_types[i]; i++) {
				if (!strcmp(supported_types[i], type)) {
					logit(LOG_DEBUG, "M-SEARCH * ST: %s from %s port %d", type,
					      inet_ntoa(sin->sin_addr), ntohs(sin->sin_port));
					send_message(ifs, type, &sa, salen);
					return;
				}
			}

			logit(LOG_DEBUG, "M-SEARCH * for unsupported ST: %s from %s", type,
			      inet_ntoa(sin->sin_addr));
		}
	}
}

static void wait_message(time_t tmo)
{
	int timeout = tmo - time(NULL);

	do {
		if (timeout < 0)
			break;

		if (ssdp_poll(timeout * 1000) == -1) {
			if (errno != EINTR)
				err(1, "Unrecoverable error");
			break;
		}

		timeout = tmo - time(NULL);

	} while (timeout);
}

static void announce(struct ifsock *ifs, int mod)
{
	if (mod && !ifs->mod)
		return;
	ifs->mod = 0;

	for (size_t i = 0; supported_types[i]; i++) {
		/* UUID sent in SSDP_ST_ALL, first announce */
		if (!strcmp(supported_types[i], uuid))
			continue;

		send_message(ifs, supported_types[i], NULL, 0);
	}
}

void ssdpd_refresh(void)
{
	if (ssdp_init(ttl, 1, ifs, ifnum, ssdp_recv) > 0) {
		logit(LOG_INFO, "Sending SSDP NOTIFY on new interfaces ...");
		ssdp_foreach(announce, 1);
	}
}

static void drop_privs(void)
{
	struct passwd *pw;

	if (chdir("/"))
		logit(LOG_WARNING, "Failed 'cd /': %s", strerror(errno));

	if (getuid())
		return;

	pw = getpwnam("ssdp");
	if (!pw)
		pw = getpwnam("nobody");
	if (!pw) {
	fail:
		logit(LOG_WARNING, "Failed dropping root privileges: %s", strerror(errno));
		return;
	}

	logit(LOG_NOTICE, "Dropping privileges to user %s (%d:%d)", pw->pw_name, pw->pw_uid, pw->pw_gid);
	if (setgid(pw->pw_gid))
		goto fail;
	if (setuid(pw->pw_uid))
		goto fail;
}

static char *strip_quotes(char *str)
{
	char *ptr;

	if (!str)
		return NULL;
	while (*str && isspace(*str))
		str++;
	if (*str == '"')
		str++;

	ptr = str;
	while (*ptr && !isspace(*ptr) && *ptr != '"')
		ptr++;
	*ptr = 0;

	return str;
}

static int os_init(void)
{
	const char *file = "/etc/os-release";
	char line[80];
	FILE *fp;

	fp = fopen(file, "r");
	if (!fp)
		return 1;

	while (fgets(line, sizeof(line), fp)) {
		char *ptr;

		line[strlen(line) - 1] = 0;

		if (!strncmp(line, "NAME", 4) && (ptr = strchr(line, '='))) {
			logit(LOG_DEBUG, "Found NAME:%s", ptr + 1);
			if (os)
				free(os);
			os = strdup(strip_quotes(++ptr));
		}

		if (!strncmp(line, "VERSION_ID", 10) && (ptr = strchr(line, '='))) {
			logit(LOG_DEBUG, "Found VERSION_ID:%s", ptr + 1);
			if (ver)
				free(ver);
			ver = strdup(strip_quotes(++ptr));
		}

		if (!strncmp(line, "VERSION_CODENAME", 16) && (ptr = strchr(line, '='))) {
			logit(LOG_DEBUG, "Found VERSION_CODENAME:%s", ptr + 1);
			if (!ver)
				ver = strdup(strip_quotes(++ptr));
		}
	}

	logit(LOG_DEBUG, "Found os:%s ver:%s", os, ver);
	fclose(fp);

	return 0;
}

static void lsb_init(void)
{
	const char *file = "/etc/lsb-release";
	int severity = LOG_INFO;
	char line[80];
	char *ptr;
	FILE *fp;

	if (!os_init())
		goto fallback;

	fp = fopen(file, "r");
	if (!fp) {
#ifndef __linux__
		struct utsname uts;

		if (!uname(&uts)) {
			os  = strdup(uts.sysname);
			ver = strdup(uts.release);
		}
#endif
		goto fallback;
	}

	while (fgets(line, sizeof(line), fp)) {
		line[strlen(line) - 1] = 0;

		ptr = strstr(line, "DISTRIB_ID");
		if (ptr && (ptr = strchr(ptr, '='))) {
			if (os)
				free(os);
			os = strdup(++ptr);
		}

		ptr = strstr(line, "DISTRIB_RELEASE");
		if (ptr && (ptr = strchr(ptr, '='))) {
			if (ver)
				free(ver);
			ver = strdup(++ptr);
		}
	}
	fclose(fp);

fallback:
	if (os && ver)
		snprintf(server_string, sizeof(server_string), "%s/%s UPnP/1.0 %s/%s",
			 os, ver, PACKAGE_NAME, PACKAGE_VERSION);
	else {
		logit(LOG_WARNING, "No %s found on system, using built-in server string.", file);
		severity = LOG_WARNING;
	}

	logit(severity, "Server: %s", server_string);
}

static void lsb_exit(void)
{
	if (os)
		free(os);
	if (ver)
		free(ver);
}

/*
 * _CACHEDIR is the configurable fallback.  We only read that, if it
 * exists, otherwise we use the system _PATH_VARDB, which works on all
 * *BSD and GLIBC based Linux systems.  Some Linux systms don't have the
 * correct FHS /var/lib/misc for that define, so we check for that too.
 */
static FILE *fopen_cache(char *mode, char *fn, size_t len)
{
	FILE *fp;

	if (cachefn) {
		strlcpy(fn, cachefn, len);
		return fopen(fn, mode);
	}

	snprintf(fn, len, _CACHEDIR "/" PACKAGE_NAME ".cache");
	fp = fopen(fn, mode);
	if (!fp) {
		snprintf(fn, len, "%s/" PACKAGE_NAME ".cache", _PATH_VARDB);
		fp = fopen(fn, mode);

		/* fallback if /var/lib/misc doesn't exist */
		if (!fp && errno == ENOENT && mode[0] == 'w') {
			(void)mkdir("/var/lib/" PACKAGE_NAME, 0755);
			snprintf(fn, len, "/var/lib/%s/uuid.cache", PACKAGE_NAME);
			fp = fopen(fn, mode);
		}
	}

	return fp;
}

/* https://en.wikipedia.org/wiki/Universally_unique_identifier */
static void uuidgen(void)
{
	char file[256];
	char buf[42];
	FILE *fp;

	if (uuid[0])
		goto custom;

	fp = fopen_cache("r", file, sizeof(file));
	if (!fp) {
	generate:
		fp = fopen_cache("w", file, sizeof(file));
		if (!fp)
			logit(LOG_WARNING, "Cannot create UUID cache, %s: %s", file, strerror(errno));
		else
			logit(LOG_DEBUG, "Opened cache file %s for saving UUID", file);

		srand(time(NULL));
		snprintf(buf, sizeof(buf), "uuid:%8.8x-%4.4x-%4.4x-%4.4x-%6.6x%6.6x",
			 rand() & 0xFFFFFFFF,
			 rand() & 0xFFFF,
			 (rand() & 0x0FFF) | 0x4000, /* M  4 MSB version => version 4 */
			 (rand() & 0x1FFF) | 0x8000, /* N: 3 MSB variant => variant 1 */
			 rand() & 0xFFFFFF, rand() & 0xFFFFFF);

		if (fp) {
			logit(LOG_DEBUG, "Creating new UUID cache file, %s", file);
			fprintf(fp, "%s\n", buf);
			fclose(fp);
		}
	} else {
		logit(LOG_DEBUG, "Opened UUID cache file %s", file);
		if (!fgets(buf, sizeof(buf), fp)) {
			fclose(fp);
			goto generate;
		}
		if (buf[strlen(buf) - 1] == '\n')
			buf[strlen(buf) - 1] = 0;
		fclose(fp);
	}

	strlcpy(uuid, buf, sizeof(uuid));
custom:
	logit(LOG_DEBUG, "URN: %s", uuid);
	if (cachefn) {
		free(cachefn);
		cachefn = NULL;
	}
}

static void sighandler(int signo)
{
	switch (signo) {
	case SIGALRM:
		recheck = 1;
		break;
	default:
		running = 0;
		break;
	}
}

static void signal_init(void)
{
	signal(SIGTERM, sighandler);
	signal(SIGINT,  sighandler);
	signal(SIGHUP,  sighandler);
	signal(SIGQUIT, sighandler);
	signal(SIGPIPE, SIG_IGN); /* get EPIPE instead */
	signal(SIGALRM, sighandler);
}

static int usage(int code)
{
	printf("Usage: %s [-hnsvw] [-c FILE] [-d URL] [-i SEC] [-l LEVEL] [-m NAME] [-M URL]\n"
	       "                      [-p URL] [-P FILE] [-r SEC] [-R NUM] [-t TTL] [-u UUID]\n"
	       "                      [IFACE [IFACE ...]]\n"
	       "\n"
	       "    -c FILE   Path to alternate ssdpd.cache to store and/or read the UUID\n"
	       "    -d URL    Override UPnP description.xml URL in announcements.  The '%%s' in\n"
	       "              the URL is replaced with the IP, e.g. https://%%s:1901/main.xml\n"
	       "    -h        This help text\n"
	       "    -i SEC    SSDP notify interval (30-900), default %d sec\n"
	       "    -l LVL    Set log level: none, err, notice (default), info, debug\n"
	       "    -m NAME   Override manufacturer in the default description.xml\n"
	       "    -M URL    Override manufacturerURL in the default description.xml\n"
	       "    -n        Run in foreground, do not daemonize by default\n"
	       "    -r SEC    Interface refresh interval (5-1800), default %d sec\n"
	       "    -R NUM    Initial retries, using 10 sec refresh interval, default 3 times\n"
	       "              Note: unused on systems with netlink interface monitoring.\n"
	       "    -p URL    Override presentationURL (WebUI) in the default description.xml\n"
	       "              The '%%s' is replaced with the IP address.  Default: http://%%s/\n"
	       "    -P FILE   Override PID file location, absolute path required\n"
	       "    -s        Use syslog, default unless running in foreground, -n\n"
	       "    -t TTL    TTL for multicast frames, default 2, according to the UDA\n"
	       "    -u UUID   Custom UUID instead of auto-generating one\n"
	       "    -v        Show program version\n"
	       "    -w        Disable built-in micro HTTP server on port %d\n"
	       "\n"
	       "Bug report address : %s\n", PACKAGE_NAME, NOTIFY_INTERVAL, REFRESH_INTERVAL,
	       LOCATION_PORT, PACKAGE_BUGREPORT);
#ifdef PACKAGE_URL
        printf("Project homepage   : %s\n", PACKAGE_URL);
#endif

	return code;
}

int main(int argc, char *argv[])
{
	time_t itmo = 0;
	int background = 1;
	int interval = NOTIFY_INTERVAL;
	int refresh = REFRESH_INTERVAL;
	int initial = 10;
	int inicnt = 3;
	char *description = NULL;
	char *pidfn = PACKAGE_NAME;
	int do_syslog = 1;
	int do_web = 1;
	int nlmon = 0;
	int c;

	while ((c = getopt(argc, argv, "c:d:hi:l:m:M:np:P:r:R:st:u:vw")) != EOF) {
		switch (c) {
		case 'c':
			cachefn = strdup(optarg);
			break;

		case 'd':
			description = optarg;
			break;

		case 'h':
			return usage(0);

		case 'i':
			interval = atoi(optarg);
			if (interval < 30 || interval > 900)
				errx(1, "Invalid announcement interval (30-900).");
			break;

		case 'l':
			log_level = log_str2lvl(optarg);
			if (-1 == log_level)
				return usage(1);
			break;

		case 'm':
			strlcpy(mfrnm, optarg, sizeof(mfrnm));
			break;

		case 'M':
			strlcpy(mfrurl, optarg, sizeof(mfrurl));
			break;

		case 'n':
			background = 0;
			do_syslog--;
			break;

		case 'p':
			strlcpy(url, optarg, sizeof(url));
			break;

		case 'P':
			pidfn = optarg;
			break;

		case 'r':
			refresh = atoi(optarg);
			if (refresh < 5 || refresh > 1800)
				errx(1, "Invalid refresh interval (5-1800).");
			break;

		case 'R':
			inicnt = atoi(optarg);
			break;

		case 's':
			do_syslog++;
			break;

		case 't':
			ttl = atoi(optarg);
			if (ttl < 1 || ttl > 255)
				errx(1, "Invalid TTL (1-255).");
			break;

		case 'u':
			snprintf(uuid, sizeof(uuid), "uuid:%s", optarg);
			break;

		case 'v':
			puts(PACKAGE_VERSION);
			return 0;

		case 'w':
			do_web = 0;
			break;

		default:
			break;
		}
	}

	signal_init();

	if (background) {
		if (daemon(0, 0))
			err(1, "Failed daemonizing");
	}

	log_init(do_syslog);
	uuidgen();
	lsb_init();
	if (do_web)
		web_init();
	if (description)
		strlcpy(location, description, sizeof(location));
	else
		snprintf(location, sizeof(location), "http://%s:%d%s",
			 "%s", LOCATION_PORT, LOCATION_DESC);

	if (netlink_init() == 0)
		nlmon = 1;
	pidfile(pidfn);
	drop_privs();

	ifnum = argc - optind;
	ifs = calloc(ifnum, sizeof(char *));
	for (size_t i = 0; i < ifnum; i++)
		ifs[i] = argv[optind + i];

	while (running) {
		time_t now = time(NULL);

		if (recheck) {
			ssdpd_refresh();

			/*
			 * If we haven't got any sockets open yet, or we've
			 * just started up.  Handle things differently, we may
			 * start well before we get an initial DHCP lease.
			 */
			if (!nlmon && !ssdp_num_sockets() && inicnt > 0) {
				alarm(initial);
				inicnt--;
			} else
				alarm(refresh);

			recheck = 0;
		}

		if (itmo <= now) {
			logit(LOG_INFO, "Sending SSDP NOTIFY ...");
			ssdp_foreach(announce, 0);
			itmo = now + interval;
		}

		wait_message(itmo);
	}

	lsb_exit();
	log_exit();
	if (ifnum)
		free(ifs);

	return ssdp_exit();
}

/**
 * Local Variables:
 *  indent-tabs-mode: t
 *  c-file-style: "linux"
 * End:
 */
